SonarQube.MSBuild.Tasks/MakeUniqueDir.cs

```diff
@@ -0,0 +1,101 @@
+﻿/*
+ * SonarQube Scanner for MSBuild
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+using Mutex = System.Threading.Mutex;
+
+namespace SonarQube.MSBuild.Tasks
+{
+    public class MakeUniqueDir : Task
+    {
+        /// <summary>
+        /// The path where folder will be created, usually that is .sonarqube/out in the solution root
+        /// </summary>
+        [Required]
+        public string Path { get; set; }
+
+        /// <summary>
+        /// Optional folder name prefix
+        /// </summary>
+        public string Prefix { get; set; }
+
+        /// <summary>
+        /// Returns the name of the created folder
+        /// </summary>
+        [Output]
+        public string UniqueName { get; set; }
+
+        /// <summary>
+        /// Returns the full path of the created folder
+        /// </summary>
+        [Output]
+        public string UniquePath { get; set; }
+
+        public override bool Execute()
+        {
+            var mutex = new Mutex(false, StripReservedChars(Path));
+
+            mutex.WaitOne();
+
+            try
+            {
+                for (var i = 0; /* endless */ ; i++)
+                {
+                    var name = $"{Prefix ?? string.Empty}{i}";
+                    var fullPath = System.IO.Path.Combine(Path, name);
+                    if (!Directory.Exists(fullPath))
+                    {
+                        Directory.CreateDirectory(fullPath);
+                        UniqueName = name;
+                        UniquePath = fullPath;
+                        break;
+                    }
+                }
+            }
+            finally
+            {
+                mutex.ReleaseMutex();
```

Comment:
Shouldn't we make this class disposable and ensure we release the mutex if the class is being disposed instead of keeping it hanging?

Reply:
It should be fine to release in finally... However, if for some reason the scanner is interrupted with TaskManager for example, it is possible that the mutex enters abandoned state (e.g. when the owning thread stops without releasing), which will have impact on the next thread that tries to claim the mutex. Look for "AbandonedMutexException" in the following article:
https://msdn.microsoft.com/en-us/library/system.threading.mutex(v=vs.110).aspx

I am not sure if IDisposable will be better in this case.